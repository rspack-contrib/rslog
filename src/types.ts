export type LogLevel = 'error' | 'warn' | 'info' | 'log' | 'verbose';

export type LogMessage = number | string | Error | null;

export type LogFormatter = (
  input: string | number | null | undefined,
) => string;

export interface LogType {
  label?: string;
  level: LogLevel;
  formatter?: LogFormatter;
}

export type LogFunction = (message?: LogMessage, ...args: any[]) => void;

export interface Options {
  level?: LogLevel;
}
