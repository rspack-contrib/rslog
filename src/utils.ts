import supportsColor from 'supports-color';

// https://github.com/chalk/supports-color
export const colorLevel = supportsColor.stdout ? supportsColor.stdout.level : 0;

let errorStackRegExp = /at\s.*:\d+:\d+[\s\)]*$/;
let anonymousErrorStackRegExp = /at\s.*\(<anonymous>\)$/;

export let isErrorStackMessage = (message: string) =>
  errorStackRegExp.test(message) || anonymousErrorStackRegExp.test(message);
