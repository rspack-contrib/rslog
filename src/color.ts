import { isColorSupported } from './utils';
import type { LogFormatter } from './types';

let formatter =
  (open: string, close: string, replace = open): LogFormatter =>
  input => {
    let string = '' + input;
    let index = string.indexOf(close, open.length);
    return ~index
      ? open + replaceClose(string, close, replace, index) + close
      : open + string + close;
  };

let replaceClose = (
  string: string,
  close: string,
  replace: string,
  index: number,
): string => {
  let start = string.substring(0, index) + replace;
  let end = string.substring(index + close.length);
  let nextIndex = end.indexOf(close);
  return ~nextIndex
    ? start + replaceClose(end, close, replace, nextIndex)
    : start + end;
};

export let color = {
  bold: isColorSupported
    ? formatter('\x1b[1m', '\x1b[22m', '\x1b[22m\x1b[1m')
    : String,
  red: isColorSupported ? formatter('\x1b[31m', '\x1b[39m') : String,
  green: isColorSupported ? formatter('\x1b[32m', '\x1b[39m') : String,
  yellow: isColorSupported ? formatter('\x1b[33m', '\x1b[39m') : String,
  magenta: isColorSupported ? formatter('\x1b[35m', '\x1b[39m') : String,
  cyan: isColorSupported ? formatter('\x1b[36m', '\x1b[39m') : String,
  gray: isColorSupported ? formatter('\x1b[90m', '\x1b[39m') : String,
};
