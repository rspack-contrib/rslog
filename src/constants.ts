import { red, yellow, cyan, green, magenta } from './color';
import type { LogType } from './types';

export let LOG_LEVEL = {
  error: 0,
  warn: 1,
  info: 2,
  log: 3,
  verbose: 4,
} as const;

export let LOG_TYPES = {
  // Level error
  error: {
    label: 'error',
    level: 'error',
    color: red,
  },
  // Level warn
  warn: {
    label: 'warn',
    level: 'warn',
    color: yellow,
  },
  // Level info
  info: {
    label: 'info',
    level: 'info',
    color: cyan,
  },
  start: {
    label: 'start',
    level: 'info',
    color: cyan,
  },
  ready: {
    label: 'ready',
    level: 'info',
    color: green,
  },
  success: {
    label: 'success',
    level: 'info',
    color: green,
  },
  // Level log
  log: {
    level: 'log',
  },
  // Level debug
  debug: {
    label: 'debug',
    level: 'verbose',
    color: magenta,
  },
} satisfies Record<string, LogType>;
