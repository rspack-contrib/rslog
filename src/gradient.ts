import { bold, cyan } from './color';
import { colorLevel } from './utils';

// RGB for #bdfff3
let startColor = [189, 255, 243];
// RGB for #4ac29a
let endColor = [74, 194, 154];

let isWord = (char: string) => !/[\s\n]/.test(char);

export let gradient = (message: string) => {
  if (colorLevel < 3) {
    return colorLevel === 2 ? bold(cyan(message)) : message;
  }

  // split string and handle emoji correctly
  // https://stackoverflow.com/questions/24531751/how-can-i-split-a-string-containing-emoji-into-an-array
  let chars = [...message];
  let steps = chars.filter(isWord).length;
  let r = startColor[0];
  let g = startColor[1];
  let b = startColor[2];
  let rStep = (endColor[0] - r) / steps;
  let gStep = (endColor[1] - g) / steps;
  let bStep = (endColor[2] - b) / steps;
  let output = '';

  for (let char of chars) {
    if (isWord(char)) {
      r += rStep;
      g += gStep;
      b += bStep;
    }
    output += `\x1b[38;2;${Math.round(r)};${Math.round(g)};${Math.round(
      b,
    )}m${char}\x1b[39m`;
  }

  return bold(output);
};
