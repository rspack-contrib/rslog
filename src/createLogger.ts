import { color } from './color';
import { gradient } from './gradient';
import { LOG_LEVEL, LOG_TYPES } from './constants';
import { isErrorStackMessage } from './utils';
import type { Options, LogMessage, LogFunction } from './types';

type Keys = keyof typeof LOG_TYPES;

export let createLogger = (options: Options = {}) => {
  let maxLevel = options.level || 'log';

  let log = (type: Keys, message?: LogMessage, ...args: string[]) => {
    if (LOG_LEVEL[LOG_TYPES[type].level] > LOG_LEVEL[maxLevel]) {
      return;
    }

    if (message === undefined || message === null) {
      console.log();
      return;
    }

    let logType = LOG_TYPES[type];

    let label = '';

    if ('label' in logType) {
      label = (logType.label || '').padEnd(7);
      label = color.bold(logType.formatter ? logType.formatter(label) : label);
    }

    let text = '';

    if (message instanceof Error) {
      if (message.stack) {
        let [name, ...rest] = message.stack.split('\n');
        text = `${name.replace('Error: ', '')}\n${color.gray(rest.join('\n'))}`;
      } else {
        text = message.message;
      }
    }
    // change the color of error stacks to gray
    else if (logType.level === 'error' && typeof message === 'string') {
      let lines = message.split('\n');
      text = lines
        .map(line => (isErrorStackMessage(line) ? color.gray(line) : line))
        .join('\n');
    } else {
      text = `${message}`;
    }

    let log = label.length ? `${label} ${text}` : text;

    console.log(log, ...args);
  };

  type Logger = Record<Keys, LogFunction> & {
    greet: (message: string) => void;
  };

  let logger = {
    greet: (message: string) => log('log', gradient(message)),
  } as Logger;

  (Object.keys(LOG_TYPES) as Keys[]).forEach(key => {
    logger[key] = (...args) => log(key, ...args);
  });

  return logger;
};
