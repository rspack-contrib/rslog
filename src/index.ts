import { createLogger } from './createLogger';

export { createLogger } from './createLogger';

export let logger = createLogger();

export type {
  Options,
  LogType,
  LogLevel,
  LogMessage,
  LogFunction,
  LogFormatter,
} from './types';
