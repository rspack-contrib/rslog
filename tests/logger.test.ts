import { createLogger, logger } from '../src';
import { join } from 'path';
import { expect, test, describe, vi, Mock } from 'vitest';

const root = join(__dirname, '..');

expect.addSnapshotSerializer({
  test: val => typeof val === 'string' && val.includes(root),
  print: val => {
    return (val as any).toString().replaceAll(root, '<ROOT>');
  },
});

describe('logger', () => {
  test('should log as expected', () => {
    console.log = vi.fn();

    logger.greet(`ðŸ˜Š Rslog v1.0.0\n`);
    logger.log('this is a log message');
    logger.info('this is an info message');
    logger.warn('this is a warn message');
    logger.ready('this is a ready message');
    logger.debug('this is a debug message');
    logger.success('this is a success message');

    expect((console.log as Mock).mock.calls).toMatchSnapshot();
  });

  test('should create new logger with info level correctly', () => {
    console.log = vi.fn();

    const logger = createLogger({
      level: 'info',
    });

    logger.greet(`ðŸ˜Š Rslog v1.0.0\n`);
    logger.error('this is a error message');
    logger.info('this is an info message');
    logger.warn('this is a warn message');
    logger.ready('this is a ready message');
    logger.debug('this is a debug message');
    logger.success('this is a success message');

    expect((console.log as Mock).mock.calls).toMatchSnapshot();
  });

  test('should create new logger with warn level correctly', () => {
    console.log = vi.fn();

    const logger = createLogger({
      level: 'warn',
    });

    logger.greet(`ðŸ˜Š Rslog v1.0.0\n`);
    logger.error('this is a error message');
    logger.info('this is an info message');
    logger.warn('this is a warn message');
    logger.ready('this is a ready message');
    logger.debug('this is a debug message');
    logger.success('this is a success message');

    expect((console.log as Mock).mock.calls).toMatchSnapshot();
  });

  test('should log error with stack correctly', () => {
    console.log = vi.fn();

    logger.error(new Error('this is an error message'));

    expect((console.log as Mock).mock.calls[0][0]).toMatchSnapshot();
  });
});
