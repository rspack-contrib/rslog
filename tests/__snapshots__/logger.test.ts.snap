// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`logger > should create new logger with info level correctly 1`] = `
[
  [
    "[1m[31merror  [39m[22m this is a error message",
  ],
  [
    "[1m[36minfo   [39m[22m this is an info message",
  ],
  [
    "[1m[33mwarn   [39m[22m this is a warn message",
  ],
  [
    "[1m[32mready  [39m[22m this is a ready message",
  ],
  [
    "[1m[32msuccess[39m[22m this is a success message",
  ],
]
`;

exports[`logger > should create new logger with warn level correctly 1`] = `
[
  [
    "[1m[31merror  [39m[22m this is a error message",
  ],
  [
    "[1m[33mwarn   [39m[22m this is a warn message",
  ],
]
`;

exports[`logger > should log as expected 1`] = `
[
  [
    "[1m[38;2;179;250;236mðŸ˜Š[39m[38;2;179;250;236m [39m[38;2;170;245;228mR[39m[38;2;160;240;221ms[39m[38;2;151;235;213ml[39m[38;2;141;230;206mo[39m[38;2;131;224;199mg[39m[38;2;131;224;199m [39m[38;2;122;219;191mv[39m[38;2;112;214;184m1[39m[38;2;103;209;176m.[39m[38;2;93;204;169m0[39m[38;2;84;199;161m.[39m[38;2;74;194;154m0[39m[38;2;74;194;154m
[39m[22m",
  ],
  [
    "this is a log message",
  ],
  [
    "[1m[36minfo   [39m[22m this is an info message",
  ],
  [
    "[1m[33mwarn   [39m[22m this is a warn message",
  ],
  [
    "[1m[32mready  [39m[22m this is a ready message",
  ],
  [
    "[1m[32msuccess[39m[22m this is a success message",
  ],
]
`;

exports[`logger > should log error with stack correctly 1`] = `
[1m[31merror  [39m[22m this is an error message
[90m    at <ROOT>/tests/logger.test.ts:68:18
    at file://<ROOT>/node_modules/.pnpm/@vitest+runner@2.0.5/node_modules/@vitest/runner/dist/index.js:146:14
    at file://<ROOT>/node_modules/.pnpm/@vitest+runner@2.0.5/node_modules/@vitest/runner/dist/index.js:61:7
    at runTest (file://<ROOT>/node_modules/.pnpm/@vitest+runner@2.0.5/node_modules/@vitest/runner/dist/index.js:960:17)
    at runSuite (file://<ROOT>/node_modules/.pnpm/@vitest+runner@2.0.5/node_modules/@vitest/runner/dist/index.js:1116:15)
    at runSuite (file://<ROOT>/node_modules/.pnpm/@vitest+runner@2.0.5/node_modules/@vitest/runner/dist/index.js:1116:15)
    at runFiles (file://<ROOT>/node_modules/.pnpm/@vitest+runner@2.0.5/node_modules/@vitest/runner/dist/index.js:1173:5)
    at startTests (file://<ROOT>/node_modules/.pnpm/@vitest+runner@2.0.5/node_modules/@vitest/runner/dist/index.js:1182:3)
    at file://<ROOT>/node_modules/.pnpm/vitest@2.0.5_@types+node@16.11.7_supports-color@9.4.0/node_modules/vitest/dist/chunks/runBaseTests.CyvqmuC9.js:130:11
    at withEnv (file://<ROOT>/node_modules/.pnpm/vitest@2.0.5_@types+node@16.11.7_supports-color@9.4.0/node_modules/vitest/dist/chunks/runBaseTests.CyvqmuC9.js:94:5)[39m
`;
